File: src/utils/encryption.ts
/**
 * Encryption utility for sensitive data fields
 * Uses AES-256 encryption for encrypting storage locations and verification notes
 * Compatible with backend Fernet encryption
 */

import CryptoJS from 'crypto-js';

// Get encryption key from environment (or secure storage in production)
// In production, this should be stored securely (e.g., React Native Keychain)
const ENCRYPTION_KEY = process.env.ENCRYPTION_KEY || 'dev-key-only-change-in-prod';

/**
 * Encrypt a string using AES-256
 * @param text - The plaintext string to encrypt
 * @returns Encrypted string (base64 encoded)
 */
export const encrypt = (text: string): string => {
  if (!text) return '';
  
  try {
    const encrypted = CryptoJS.AES.encrypt(text, ENCRYPTION_KEY).toString();
    return encrypted;
  } catch (error) {
    console.error('Encryption error:', error);
    throw new Error('Failed to encrypt data');
  }
};

/**
 * Decrypt an AES-256 encrypted string
 * @param encryptedText - The encrypted string to decrypt
 * @returns Decrypted plaintext string
 */
export const decrypt = (encryptedText: string): string => {
  if (!encryptedText) return '';
  
  try {
    const decrypted = CryptoJS.AES.decrypt(encryptedText, ENCRYPTION_KEY);
    return decrypted.toString(CryptoJS.enc.Utf8);
  } catch (error) {
    console.error('Decryption error:', error);
    throw new Error('Failed to decrypt data');
  }
};

/**
 * One-way hash for comparison purposes (e.g., verifying data integrity)
 * @param text - The string to hash
 * @returns SHA-256 hash of the string
 */
export const hash = (text: string): string => {
  return CryptoJS.SHA256(text).toString();
};

File: backend/app/utils/encryption.py
"""
Encryption utility for sensitive data fields
Uses Fernet (symmetric encryption) for encrypting storage locations and verification notes
"""
import os
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC  # Note: Use PBKDF2HMAC, not PBKDF2
import base64
from app.config import settings


class EncryptionManager:
    def __init__(self):
        # Get encryption key from environment variable or settings
        key_string = os.getenv('ENCRYPTION_KEY') or getattr(settings, 'ENCRYPTION_KEY', None)
        
        if not key_string:
            raise ValueError("ENCRYPTION_KEY environment variable not set. Please set it in .env file")
        
        # Derive a proper Fernet key from the string
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=b'gharmitra_salt_2024',  # Use a fixed salt (or store securely)
            iterations=100000,
        )
        key = base64.urlsafe_b64encode(kdf.derive(key_string.encode()))
        self.cipher = Fernet(key)
    
    def encrypt(self, text: str) -> str:
        """Encrypt a string and return base64-encoded result."""
        if not text:
            return ""
        
        try:
            encrypted = self.cipher.encrypt(text.encode())
            return base64.urlsafe_b64encode(encrypted).decode()
        except Exception as e:
            print(f"Encryption error: {e}")
            raise
    
    def decrypt(self, encrypted_text: str) -> str:
        """Decrypt a base64-encoded encrypted string."""
        if not encrypted_text:
            return ""
        
        try:
            decoded = base64.urlsafe_b64decode(encrypted_text.encode())
            decrypted = self.cipher.decrypt(decoded)
            return decrypted.decode()
        except Exception as e:
            print(f"Decryption error: {e}")
            raise


# Singleton instance
_encryption_manager = None


def get_encryption_manager() -> EncryptionManager:
    """Get or create the encryption manager singleton"""
    global _encryption_manager
    if _encryption_manager is None:
        _encryption_manager = EncryptionManager()
    return _encryption_manager


def encrypt(text: str) -> str:
    """Encrypt a string using the encryption manager"""
    return get_encryption_manager().encrypt(text)


def decrypt(encrypted_text: str) -> str:
    """Decrypt an encrypted string using the encryption manager"""
    return get_encryption_manager().decrypt(encrypted_text)

